File: packages/graph/src/insightgraph_graph/schema.py
from __future__ import annotations

import logging

from insightgraph_core.ontology.loader import load_ontology
from insightgraph_core.ontology.schema import Ontology
from insightgraph_graph.connection import Neo4jConnection

logger = logging.getLogger(__name__)


def _constraint_cypher(label: str, properties: list[str]) -> str:
    """Build a CREATE CONSTRAINT IF NOT EXISTS statement."""
    prop_clause = ", ".join(f"n.{p}" for p in properties)
    constraint_name = f"{label.lower()}_{'_'.join(properties)}_unique"
    return (
        f"CREATE CONSTRAINT {constraint_name} IF NOT EXISTS "
        f"FOR (n:{label}) REQUIRE ({prop_clause}) IS UNIQUE"
    )


def _fulltext_index_cypher(label: str, properties: list[str]) -> str:
    """Build a CREATE FULLTEXT INDEX IF NOT EXISTS statement."""
    index_name = f"{label.lower()}_search"
    prop_list = ", ".join(f"n.{p}" for p in properties)
    return f"CREATE FULLTEXT INDEX {index_name} IF NOT EXISTS FOR (n:{label}) ON EACH [{prop_list}]"


async def ensure_schema(conn: Neo4jConnection, ontology: Ontology | None = None) -> None:
    """Create all Neo4j constraints and indexes defined in the ontology.

    If *ontology* is not provided it will be loaded from the bundled YAML files.
    """
    if ontology is None:
        ontology = load_ontology()

    statements: list[str] = []

    for node_def in ontology.nodes.values():
        for constraint in node_def.constraints:
            if constraint.unique:
                statements.append(_constraint_cypher(node_def.name, constraint.unique))

        for index in node_def.indexes:
            if index.fulltext:
                statements.append(_fulltext_index_cypher(node_def.name, index.fulltext))

    async with conn.session() as session:
        for stmt in statements:
            logger.debug("Executing schema DDL: %s", stmt)
            await session.run(stmt)

    logger.info("Schema ensured: %d DDL statements executed", len(statements))

File: packages/graph/src/neo4j/schema.ts
import { loadOntology } from "@insightgraph/core";
import type { Ontology } from "@insightgraph/core";
import { Neo4jConnection } from "./connection";

function constraintCypher(label: string, properties: string[]): string {
  const propClause = properties.map((p) => `n.${p}`).join(", ");
  const constraintName = `${label.toLowerCase()}_${properties.join("_")}_unique`;
  return (
    `CREATE CONSTRAINT ${constraintName} IF NOT EXISTS ` +
    `FOR (n:${label}) REQUIRE (${propClause}) IS UNIQUE`
  );
}

function fulltextIndexCypher(label: string, properties: string[]): string {
  const indexName = `${label.toLowerCase()}_search`;
  const propList = properties.map((p) => `n.${p}`).join(", ");
  return `CREATE FULLTEXT INDEX ${indexName} IF NOT EXISTS FOR (n:${label}) ON EACH [${propList}]`;
}

export async function ensureSchema(
  conn: Neo4jConnection,
  ontology?: Ontology,
): Promise<void> {
  const ont = ontology ?? loadOntology();
  const statements: string[] = [];

  for (const nodeDef of Object.values(ont.nodes)) {
    for (const constraint of nodeDef.constraints) {
      if (constraint.unique.length > 0) {
        statements.push(constraintCypher(nodeDef.name, constraint.unique));
      }
    }
    for (const index of nodeDef.indexes) {
      if (index.fulltext.length > 0) {
        statements.push(fulltextIndexCypher(nodeDef.name, index.fulltext));
      }
    }
  }

  const session = conn.session();
  try {
    for (const stmt of statements) {
      try {
        await session.run(stmt);
      } catch {
        // May fail on Community Edition (vector/fulltext indexes)
      }
    }
  } finally {
    await session.close();
  }
}
